File: UVa Online Judge/100.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <set>
#include <unordered_map>
#include <string>
using namespace std;

int collatz(long long n) {
    static unordered_map<long long, int> memo;
    if (memo.count(n))
        return memo[n];

    if (n == 1)
        return memo[n] = 1;

    if (n % 2)
        return memo[n] = 1 + collatz(3 * n + 1);
    else
        return memo[n] = 1 + collatz(n / 2);
}

vector<int> prepare() {
    vector<int> res(1000000);
    for (int i = 1; i < res.size(); i++)
        res[i] = collatz(i);
    return res;
}

bool solve() {
    int l, r;
    if (!(cin >> l >> r))
        return 0;

    cout << l << " " << r << " ";
    if (l > r)
        swap(l, r);

    static vector<int> res = prepare();
    cout << *max_element(res.begin() + l, res.begin() + r + 1) << "\n";
    return 1;
}

int main() {
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);

    ios_base::sync_with_stdio(0);
    cin.tie(0);

    while (solve());
}

File: ACMP/100.py
n = int(input())
n //= 2

count = [[0 for sum in range(9 * n + 1)] for len in range(n + 1)]
count[0][0] = 1

for len in range(1, n + 1):
    for sum in range(0, 9 * n + 1):
        for digit in range(min(10, sum + 1)):
            count[len][sum] += count[len - 1][sum - digit]

res = 0
for sum in range(0, 9 * n + 1):
    res += count[n][sum] * count[n][sum]

print(res)
